dashboard/app/models/pd/application/application_base.rb

```diff
@@ -65,6 +65,8 @@ class ApplicationBase < ApplicationRecord
     validates_inclusion_of :application_year, in: APPLICATION_YEARS
 
     # An application either has an "incomplete" or "unreviewed" state when created.
+    # The applied_at field gets set when the status becomes 'unreviewed' for the first time
+    before_save :set_applied_date, if: -> {status_changed? && ActiveRecord::Base.connection.column_exists?(:pd_applications, :applied_at)}
```

Comment:
I see what you are doing here -- it seems like you want to ship the code change ahead of the migration? while this could work fine, the practice we would normally follow is to ship the migration, and then update the code to use it. this is better because if the new behavior is broken, it is easier to quickly rollback a code change than it is to undo a migration.

my suggestion would be to make this PR dependent on the migration, remove the `column_exists?` check, and then wait to merge this PR until the migration PR has already shipped.

Reply:
same comment below, anywhere we are using `column_exists?`.

Reply:
Oh I got different advice when I first did a migration. What happens once the migration happens, but new applications are submitted? I feel like we'll miss those applications. Are you saying I'd merge the migration, then merge this, then fix the applications that were submitted in the middle on prod?

Reply:
If the new behavior is broken, could I not revert this? I'm not sure I totally understand, but your way is certainly fewer PRs total which is nice

Reply:
Ah it looks like people aren't submitted apps right now? I just ran `Pd::Application::ApplicationBase.where(created_at: Date.today).pluck(:id)` so not sure we'd even need to worry about missing some ...

Reply:
Hmm not convinced that query works ... but it'd be a small # of apps regardless

Reply:
oh interesting, good point.  I would usually start from the long version, and then see if any steps can be combined / skipped. the long version for a generically moving from one data column to another with zero downtime is:
1. create the new column
2. update the code to start writing to the new column, while still writing to and reading from the old column
3. backfill any blank rows in the new column, using the existing values in the old column
4. manually inspect the new data to make sure it looks right (optional / discretionary)
5. switch to reading the new column, still writing to both columns
6. stop writing to the old column (not applicable in this case)
7. remove the old column (not applicable in this case)

in your case, I think this would end up doing steps 1, 2, 3, 5 as individual steps (using your judgment on whether to include step 4). 
 
the upside of this approach is that it is very safe and simple -- when the new behavior is enabled in step 5, it is easy to rollback to the old behavior by reverting the code change in step 5. another implicit advantage of this approach is that you do not have to do very much thinking -- this approach is tried and true. 

to evaluate the safety of your approach, on the other hand, we must do some thinking :-) I would ask the following questions:
1. what will happen while your migration is running, after it has created the column and before the data has been copied over? this could be a short period of time, if all goes smoothly. but, if your migration fails part way through filling the new column, then you might end up in a state where the column exists but data has not been copied over. if the result is just that a field in the workshop dashboard is wrong until the problem is fixed, then this may not so bad. but if there is data loss, then this might be considered unacceptable.
2. what happens if the migration succeeds, but you realize that the new behavior is wrong, and old behavior needs to be restored? can you simply revert this PR (the code change)? it seems to me like you can -- your new column might end up with inconsistent data, so there would be some cleanup steps, but you'd be safely back onto your old implementation with known good behavior.

at any point, if these questions get too hard or annoying to answer, you can go back to doing things the long way (steps 1-5 above).

want to take a crack at answering the above questions and meet about this tomorrow morning?

cc @bethanyaconnor since we just had a tangentially related conversation.

Reply:
> Ah it looks like people aren't submitted apps right now? 

ok, if we can confirm this, we can certainly skip / combine some steps! 

> I just ran Pd::Application::ApplicationBase.where(created_at: Date.today).pluck(:id) so not sure we'd even need to worry about missing some ...

> Hmm not convinced that query works ... but it'd be a small # of apps regardless

I am definitely seeing some new apps each day:
```
[production] dashboard > Pd::Application::TeacherApplication.count
=> 24769
[production] dashboard > puts Pd::Application::TeacherApplication.pluck(:created_at).join("\n")
...
2022-03-14 23:25:57 UTC
2022-03-15 02:18:55 UTC
2022-03-15 03:38:15 UTC
2022-03-15 12:34:30 UTC
2022-03-15 15:03:38 UTC
2022-03-15 15:08:17 UTC
2022-03-15 15:29:24 UTC
2022-03-15 15:40:53 UTC
2022-03-15 17:43:02 UTC
2022-03-15 18:14:52 UTC
2022-03-15 19:00:34 UTC
2022-03-15 21:22:49 UTC
2022-03-15 22:05:38 UTC
2022-03-16 12:19:25 UTC
2022-03-16 13:28:22 UTC
2022-03-16 14:25:24 UTC
2022-03-16 14:33:09 UTC
2022-03-16 14:45:27 UTC
2022-03-16 15:39:39 UTC
2022-03-16 16:19:48 UTC
2022-03-16 18:20:23 UTC
2022-03-16 18:37:18 UTC
2022-03-16 18:53:14 UTC
2022-03-16 19:43:45 UTC
2022-03-16 20:24:57 UTC
2022-03-16 21:03:29 UTC
```
so I think I would at least like to understand what would happen to the applications if they were submitted during a time when the new column existed but was not populated.

not trying to make this into a bigger deal than it is... the situation might be recoverable. but if the answer is that we'd lose that application irrecoverably then I think that would be a red flag.

Reply:
I am trying to find where the approach I listed is documented. sometimes it is called a "blue / green" deployment. these are what I've found so far:
* some random quora answer: https://www.quora.com/How-do-you-migrate-database-data-without-any-downtime
* some random medium post: https://engineering.wework.com/zero-downtime-deployments-with-data-migrations-2e337df48b8f

again, you don't have to do it that way, but it is a standard pattern and i hope these explanations can articulate it a bit better than I did.